python/pyspark/sql/dataframe.py

```diff
@@ -3928,6 +3928,71 @@ def dropDuplicates(self, subset: Optional[List[str]] = None) -> "DataFrame":
             jdf = self._jdf.dropDuplicates(self._jseq(subset))
         return DataFrame(jdf, self.sparkSession)
 
+    def dropDuplicatesWithinWatermark(self, subset: Optional[List[str]] = None) -> "DataFrame":
+        """Return a new :class:`DataFrame` with duplicate rows removed,
+         optionally only considering certain columns, within watermark.
+
+        For a static batch :class:`DataFrame`, it just drops duplicate rows. For a streaming
+        :class:`DataFrame`, this will keep all data across triggers as intermediate state to drop
+        duplicated rows. The state will be kept to guarantee the semantic, "Events are deduplicated
+        as long as the time distance of earliest and latest events are smaller than the delay
+        threshold of watermark." The watermark for the input :class:`DataFrame` must be set via
+        :func:`withWatermark`. Users are encouraged to set the delay threshold of watermark longer
```

Comment:
batch DataFrame does not require watermark, and we actually remove withWatermark if the query is batch one. If we require batch query to provide watermark, that will be very odd because we ignore the delay threshold in any way.

So I think there are two choices, do not support batch query since it's confusing vs just do the same with existing API for batch case.
cc. @rangadi

Reply:
I think it better to include watermark since this serves as important documentation for the users. Most of the time, the user is looking for an example to use in streaming. Not including watermark here is going to be confusing for them. 
The fact that it is ignored in batch is fine.

Reply:
I actually intend to only provide streaming examples to avoid the additional explanation on batch case. For batch, there are a bunch of tools to perform deduplication, distinct / dropDuplicates / dropDuplicatesWithinWatermark. Most of batch use cases don't need to come up with using dropDuplicatesWithinWatermark.

Reply:
@zsxwing Does my comment for leaving this as it is make sense to you? Could you please have another round of review?

Reply:
It sounds weird to me that a method is called dropDuplicates**WithinWatermark** but I don't need to set the watermark.

> For batch, there are a bunch of tools to perform deduplication, distinct / dropDuplicates / dropDuplicatesWithinWatermark. Most of batch use cases don't need to come up with using dropDuplicatesWithinWatermark.

I think the common use case for `dropDuplicatesWithinWatermark` in batch is: develop the code in batch mode and switch to streaming later. In this case, catching potential issues in batch mode is better.

Reply:
I just had a discussion with @zsxwing offline. There was a confusion that we guarantee the same output between batch and streaming for new API which isn't true. To remove any confusion from users, we agreed to remove supporting batch query. I'll reflect the decision.

Reply:
@HeartSaVioR Thanks! We can wait for the user's feedback first before supporting batch queries. Changing from an error to no error is always easier.

Reply:
This commit addressed the rollback of supporting batch query.
0608889e4d1afc4bb5d1710eef45cf50d3c29a0f

@zsxwing @rangadi Please have a quick look at the change. Thanks!

Reply:
dropDuplicates does not support exact same output between batch and streaming either. No stateful operation guarantees in the precense of late records. What is the difference here? Better to support batch in the same manner as dropDuplicates(). 
I don't think it is a good UX for customer to get errors then we fix it by relaxing. 
But I will leave the decision to you.

Reply:
If you don't have any late record then dropDuplicates() guarantees the same result. That enables users to test their code with batch query first, and then migrate to streaming query later.

dropDuplicatesWithinWatermark doesn't guarantee such thing so it's making no sense to test with batch query and migrate to streaming. That was why I had a offline discussion with @zsxwing .

Reply:
If we want to support this API in batch query, I think we have to implement the same behavior, not forwarding to dropDuplicates(). But that's also very odd because we are telling users that watermark is no-op in batch query and now we have to get delay threshold from withWatermark. I'd say it conflicts with base concept.

Reply:
(If we want to implement the same behavior to batch query, we will have to kick the part of "best effort" out as well in streaming. e.g. We deduplicate the event whenever there is an existing state, which does not say they're within delay threshold. We evict the state at the end of processing, hence we are accepting slightly more events to be deduplicated. That might be better behavior for streaming, but if we want to match the behavior between batch and streaming, the behavior must be deterministic.)